// firebase-tools/lib/apphosting/localbuilds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runUniversalMaker = runUniversalMaker;
exports.localBuild = localBuild;
const childProcess = require("child_process");
const fs = require("fs-extra");
const path = require("path");
const index_1 = require("./secrets/index");
const prompt_1 = require("../prompt");
const error_1 = require("../error");
const logger_1 = require("../logger");
const utils_1 = require("../utils");
const universalMakerDownload_1 = require("./universalMakerDownload");
async function runUniversalMaker(projectRoot, framework) {
    const universalMakerBinary = await (0, universalMakerDownload_1.getOrDownloadUniversalMaker)();
    executeUniversalMakerBinary(universalMakerBinary, projectRoot);
    return processUniversalMakerOutput(projectRoot, framework);
}
function executeUniversalMakerBinary(universalMakerBinary, projectRoot) {
    try {
        const bundleOutput = path.join(projectRoot, "bundle_output");
        fs.removeSync(bundleOutput);
        fs.ensureDirSync(bundleOutput);
        const res = childProcess.spawnSync(universalMakerBinary, ["-application_dir", projectRoot, "-output_dir", projectRoot, "-output_format", "json"], {
            env: {
                ...process.env,
                X_GOOGLE_TARGET_PLATFORM: "fah",
                FIREBASE_OUTPUT_BUNDLE_DIR: bundleOutput,
            },
            stdio: "pipe",
        });
        if (res.stdout) {
            logger_1.logger.debug("[Universal Maker stdout]:\n" + res.stdout.toString());
        }
        if (res.stderr) {
            logger_1.logger.debug("[Universal Maker stderr]:\n" + res.stderr.toString());
        }
        if (res.error) {
            throw res.error;
        }
        if (res.status !== 0) {
            throw new error_1.FirebaseError(`Universal Maker failed with exit code ${res.status ?? "unknown"}.`);
        }
    }
    catch (e) {
        if (e && typeof e === "object" && "code" in e && e.code === "EACCES") {
            throw new error_1.FirebaseError(`Failed to execute the Universal Maker binary at ${universalMakerBinary} due to permission constraints. Please assure you have set execution permissions (e.g., chmod +x) on the file.`);
        }
        throw e;
    }
}
function parseBundleYaml(projectRoot, defaultRunCommand) {
    const bundleYamlPath = path.join(projectRoot, ".apphosting", "bundle.yaml");
    if (!fs.existsSync(bundleYamlPath)) {
        throw new error_1.FirebaseError("Failed to resolve build artifacts. Ensure Universal Maker produced a valid bundle.yaml with outputFiles.");
    }
    const bundleRaw = fs.readFileSync(bundleYamlPath, "utf-8");
    const bundleData = (0, utils_1.wrappedSafeLoad)(bundleRaw);
    const runCommand = bundleData?.runConfig?.runCommand ?? defaultRunCommand;
    const outputFiles = bundleData?.outputFiles?.serverApp?.include;
    if (!outputFiles) {
        throw new error_1.FirebaseError("Failed to resolve build artifacts. Ensure Universal Maker produced a valid bundle.yaml with outputFiles.");
    }
    return { runCommand, outputFiles };
}
function processUniversalMakerOutput(projectRoot, framework) {
    const outputFilePath = path.join(projectRoot, "build_output.json");
    if (!fs.existsSync(outputFilePath)) {
        throw new error_1.FirebaseError(`Universal Maker did not produce the expected output file at ${outputFilePath}`);
    }
    const outputRaw = fs.readFileSync(outputFilePath, "utf-8");
    fs.unlinkSync(outputFilePath);
    const bundleOutput = path.join(projectRoot, "bundle_output");
    const targetAppHosting = path.join(projectRoot, ".apphosting");
    if (fs.existsSync(bundleOutput)) {
        fs.ensureDirSync(targetAppHosting);
        const files = fs.readdirSync(bundleOutput);
        for (const file of files) {
            const dest = path.join(targetAppHosting, file);
            if (fs.existsSync(dest)) {
                fs.removeSync(dest);
            }
            fs.moveSync(path.join(bundleOutput, file), dest);
        }
        fs.removeSync(bundleOutput);
    }
    let umOutput;
    try {
        umOutput = JSON.parse(outputRaw);
    }
    catch (e) {
        throw new error_1.FirebaseError(`Failed to parse build_output.json: ${(0, error_1.getErrMsg)(e)}`);
    }
    const defaultRunCommand = `${umOutput.command} ${umOutput.args.join(" ")}`;
    const { runCommand: finalRunCommand, outputFiles: finalOutputFiles } = parseBundleYaml(projectRoot, defaultRunCommand);
    return {
        metadata: {
            language: umOutput.language,
            runtime: umOutput.runtime,
            framework: framework || "nextjs",
        },
        runConfig: {
            runCommand: finalRunCommand,
            environmentVariables: Object.entries(umOutput.envVars || {})
                .filter(([k]) => k !== "FIREBASE_OUTPUT_BUNDLE_DIR")
                .map(([k, v]) => ({
                variable: k,
                value: String(v),
                availability: ["RUNTIME"],
            })),
        },
        outputFiles: {
            serverApp: {
                include: finalOutputFiles,
            },
        },
    };
}
async function localBuild(projectId, projectRoot, framework, env = {}, options) {
    const hasBuildAvailableSecrets = Object.values(env).some((v) => v.secret && (!v.availability || v.availability.includes("BUILD")));
    if (hasBuildAvailableSecrets && !options?.allowLocalBuildSecrets) {
        if (options?.nonInteractive) {
            throw new error_1.FirebaseError("Using build-available secrets during a local build in non-interactive mode requires the --allow-local-build-secrets flag.");
        }
        if (!(await (0, prompt_1.confirm)({
            message: "Your build includes secrets that are available to the build environment. Using secrets in local builds may leave sensitive values in local artifacts/temporary files. Do you want to continue?",
            default: false,
        }))) {
            throw new error_1.FirebaseError("Cancelled local build due to BUILD-available secrets.");
        }
    }
    const originalEnv = { ...process.env };
    const addedEnv = await toProcessEnv(projectId, env);
    for (const [key, value] of Object.entries(addedEnv)) {
        process.env[key] = value;
    }
    let apphostingBuildOutput;
    try {
        apphostingBuildOutput = await runUniversalMaker(projectRoot, framework);
    }
    finally {
        for (const key in process.env) {
            if (!(key in originalEnv)) {
                delete process.env[key];
            }
        }
        for (const [key, value] of Object.entries(originalEnv)) {
            process.env[key] = value;
        }
    }
    const annotations = Object.fromEntries(Object.entries(apphostingBuildOutput.metadata).map(([key, value]) => [key, String(value)]));
    const discoveredEnv = apphostingBuildOutput.runConfig.environmentVariables?.map(({ variable, value, availability }) => ({
        variable,
        value,
        availability: availability,
    }));
    return {
        outputFiles: apphostingBuildOutput.outputFiles?.serverApp.include ?? [],
        annotations,
        buildConfig: {
            runCommand: apphostingBuildOutput.runConfig.runCommand,
            env: discoveredEnv ?? [],
        },
    };
}
async function toProcessEnv(projectId, env) {
    const entries = await Promise.all(Object.entries(env).map(async ([key, value]) => {
        if (value.availability && !value.availability.includes("BUILD")) {
            return null;
        }
        if (value.secret) {
            const resolvedValue = await (0, index_1.loadSecret)(projectId, value.secret);
            return [key, resolvedValue];
        }
        else {
            return [key, value.value || ""];
        }
    }));
    const filteredEntries = entries.filter((entry) => entry !== null);
    return Object.fromEntries(filteredEntries);
}
