// firebase-tools/lib/downloadUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadToTmp = downloadToTmp;
exports.validateSize = validateSize;
exports.validateChecksum = validateChecksum;
const url_1 = require("url");
const crypto = require("crypto");
const fs = require("fs-extra");
const ProgressBar = require("progress");
const tmp = require("tmp");
const apiv2_1 = require("./apiv2");
const error_1 = require("./error");
async function downloadToTmp(remoteUrl, auth = false) {
    const u = new url_1.URL(remoteUrl);
    const c = new apiv2_1.Client({ urlPrefix: u.origin, auth });
    const tmpfile = tmp.fileSync();
    const writeStream = fs.createWriteStream(tmpfile.name);
    const res = await c.request({
        ignoreQuotaProject: true,
        method: "GET",
        path: u.pathname,
        queryParams: u.searchParams,
        responseType: "stream",
        resolveOnHTTPError: true,
    });
    if (res.status !== 200) {
        throw new error_1.FirebaseError(`download failed, status ${res.status}: ${await res.response.text()}`, {
            status: res.status,
        });
    }
    const total = parseInt(res.response.headers.get("content-length") || "0", 10);
    const totalMb = Math.ceil(total / 1000000);
    const bar = new ProgressBar(`Progress: :bar (:percent of ${totalMb}MB)`, { total, head: ">" });
    res.body.on("data", (chunk) => {
        bar.tick(chunk.length);
    });
    await new Promise((resolve) => {
        writeStream.on("finish", () => resolve());
        res.body.pipe(writeStream);
    });
    return tmpfile.name;
}
function validateSize(filepath, expectedSize) {
    return new Promise((resolve, reject) => {
        const stat = fs.statSync(filepath);
        return stat.size === expectedSize
            ? resolve()
            : reject(new error_1.FirebaseError(`download failed, expected ${expectedSize} bytes but got ${stat.size}`, { exit: 1 }));
    });
}
function validateChecksum(filepath, expectedChecksum, algorithm = "md5") {
    return new Promise((resolve, reject) => {
        const hash = crypto.createHash(algorithm);
        const stream = fs.createReadStream(filepath);
        stream.on("data", (data) => hash.update(data));
        stream.on("end", () => {
            const checksum = hash.digest("hex");
            return checksum === expectedChecksum
                ? resolve()
                : reject(new error_1.FirebaseError(`download failed, expected checksum ${expectedChecksum} but got ${checksum}`, { exit: 1 }));
        });
    });
}
